### PR TITLE
Read the memory pipeline's diagnostics through the contract

### PR DESCRIPTION
**Pull request by YellowSnnowmann**
## Summary

- `pipeline_status` and `backfill_status` read their numbers from the bound memory driver instead of `SELECT`s against TinyCortex's tables — the largest read-only group of direct engine references in #5560.
- The raw-connection door leaves `memory/tree/tree/rpc.rs`: `rusqlite::OptionalExtension`, `queue::store` and `JobStatus` all fall out.
- Fixes four real defects of the same shape: numbers that must agree being read at different instants.
- Advances the vendored engine to v1.2.0 (its own commit — see the caveat below).
- Also moves five `ensure_reembed_backfill` calls onto `Maintenance::reembed`, which turned out to need no upstream work at all.

## Problem

Seventeen sites across eight files reached past the contract into `store::chunks::store::with_connection` and queried the engine's schema directly. Every one is a read-only aggregate — counts, timestamps, one failure row — so none of it needs a SQLite handle; it needed methods that did not exist. tinymemory#85 adds them.

Writing the callers surfaced four defects in the code being replaced:

| Defect | Consequence |
| --- | --- |
| Five separate job-counter reads | `failed_unrecoverable` is a subset of `failed`; a retry between reads reports more unrecoverable failures than failures |
| Chunk count and extracted count read separately | numerator above denominator — extraction coverage over 100% |
| `queue_idle_ms` re-read the eligible count and its reference timestamps | a settle in the gap manufactures an idle window that never existed |
| The failure and its success watermark | read on one connection **deliberately** (the #5427 review caught that race) — two bus calls would have put it back |

The last one shaped the upstream API: `QueueFailure` carries `last_success_ms`, read alongside the failure, so the caller still gets one observation.

## Solution

Three `Maintenance` methods answer all four sites. Two host functions became pure as a result — `queue_idle_ms` derives from a snapshot rather than querying, and the supersession rule splits out of the fetch as `blocking_cause`. Those are the parts with edge cases (an untimestamped failure, a watermark on the same millisecond, a reason this build has no code for), and all three now have tests they could not have when reaching them meant standing up a store.

**Testing trade, worth knowing before the next migration.** A handler reading through the contract cannot be proven by writing rows behind it: the real driver is a compiled module a unit test cannot load, so a test workspace binds the null driver and every diagnostic reads empty. `binding::FixedDiagnostics` + `install_diagnostics_for_test` put a driver in between, and the tests split along the seam — what the host *derives* stays here; what a store *is* (an ingest raising a count, a deferred job staying ready without becoming eligible) moves to the driver's conformance suite, where a real store exists. Three `read_rpc_tests` needed the same treatment without naming a migrated handler, because `vault_health_check_rpc` folds `pipeline_status_rpc` in.

**Two behaviours preserved deliberately rather than improved in passing.** An empty store still reports `Some(0.0)` coverage — `None` has always meant "unavailable", and whether an empty store should read as 0% is a decision about what the panel shows, not a consequence of moving a read. And `doctor` still counts rows itself: `store_stats` answers exactly its question but is `async`, and that module's doc is explicit that `run` stays blocking-only with async probes hoisted into the caller.

**This sheds no dependency on its own.** `tinymemory-core` stays linked while chunk reads, the re-embed queue and the ingest pipeline still name it from the same file. #5560's payoff is all-or-nothing; the allowlist entry records what left and what did not.

### ⚠️ The engine bump needs a decision before this merges

`vendor/tinymemory` was 146 commits behind and the tree **did not compile** against v1.2.0: `RecallOpts` gained `exclude_session_id`. It is set to `None` in its own commit, marked TODO, and that is bug-compatible rather than correct — the engine's own comment says the ambient task-local it replaces reads `None` across the module boundary (a `cdylib` has its own statics), and this call goes through the module, so `None` hands the agent back what it just said. The engine's accessor for the ambient value is `pub(crate)`, so the host cannot simply forward it.

That commit is deliberately separable. Reviewing 146 commits of engine alongside a host change reviews neither, and it may be better as its own PR with this rebased on top.

## The re-embed migration, and what it says about the audit

Five sites called `tinymemory_core::queue::ensure_reembed_backfill` after an embedder-adjacent save. `MemoryMaintenance::reembed` is documented as "recompute embeddings for content whose embedding is missing or stale", and the TinyCortex driver implements it by calling `ensure_reembed_backfill` and reporting how many jobs it enqueued — the same operation, reachable through the contract the whole time.

The audit had recorded the re-embed queue as having "no capability family". That was true of the requeue half and wrong about the enqueue half; nobody had checked what the driver already did with the nearest existing method. The module docs now say to check that first, because it is the cheapest thing to rule out before sizing an upstream ask.

Two test consequences worth knowing before the next migration:

- **A handler that resolves a binding pays for one when nothing installed it.** `store_session_defers_live_jwt_when_auth_me_hangs_past_budget` asserts a 200ms validation budget and started taking **4.4 seconds** — resolving the driver means attempting to load the compiled module, which a unit test cannot do but takes seconds to fail at. Binding a driver in that file's `test_config` took the whole suite from **97s to 45s**, so other tests were paying it silently.
- **Row-counting tests become ask-recording tests.** `store_session_requeues_reembed_backfill_after_login` counted `mem_tree_jobs`; the row is the driver's doing now, so it records that the host asked — once, since a second ask would enqueue a second chain over the same uncovered rows.

**No file leaves the ratchet.** `config/ops/model.rs` and `inference/embeddings/rpc.rs` still call `requeue_failed_after_provider_change`, which genuinely has no family; the other three have unrelated blockers. Their allowlist reasons say so now rather than naming a call that has moved.

## Submission Checklist

- [x] Tests added or updated (happy path + at least one failure / edge case) — new coverage for the supersession edge cases (same-millisecond settle, untimestamped failure), the backlog-vs-stall derivation, and the driver's aggregates reaching the wire
- [x] **Diff coverage ≥ 80%** — full suite green: `11557 passed; 0 failed` (`cargo test --lib` with the product feature set)
- [x] N/A: behaviour-preserving refactor — Coverage matrix updated: no feature rows added, removed or renamed
- [x] No new external network dependencies introduced
- [x] N/A: no release-cut surface changes — Manual smoke checklist: the memory status panel renders the same fields
- [x] N/A: deliberately not closing — Linked issue: this is one group of #5560, not the whole issue, so it stays open

## Impact

Desktop/CLI. No wire-shape change — `PipelineStatusResponse` and `BackfillStatusResponse` carry the same fields with the same meanings. Behaviour differences are the four defect fixes above, all in the direction of reporting fewer false alarms: a fast-failing queue no longer reads as stalled, coverage cannot exceed 100%, and counter sets stay self-consistent.

One performance note in the other direction: each handler resolves the binding (a cached lookup) and the driver does its own `spawn_blocking`, so the host's own `spawn_blocking` wrappers are gone. Round trips did not increase — five queue reads became one, three chunk reads became one.

## Related

Builds on tinyhumansai/tinymemory#85 (merged) — the `Maintenance` methods this calls. The engine pin names upstream `main`.

Part of #5560.

## Validation

```
cargo test --lib --features "$(bash scripts/ci/product-features.sh)"   # 11559 passed; 0 failed
cargo clippy --lib --features "$(bash scripts/ci/product-features.sh)" -- -D warnings   # clean
cargo fmt --all -- --check                                             # clean
```

## Summary by CodeRabbit

- **New Features**
  - Improved memory pipeline and backfill status reporting using current provider diagnostics.
  - Added support for retrying failed memory jobs and re-embedding content after relevant settings, credential, or synchronization changes.
  - Enhanced memory filtering to respect excluded sessions for threaded and unthreaded recall.
  - Improved ingest error classification and chunk-list filtering.

- **Bug Fixes**
  - Improved reliability of memory health, pipeline, and maintenance operations when optional capabilities are unavailable.
  - Updated memory tooling and tests to use the active storage configuration consistently.

### CODE REVIEW

File: src/openhuman/agent/tinyagents/host/agent_memory.rs [hunk 1]
```diff
@@ -365,6 +365,15 @@ impl AgentMemory for OpenHumanAgentMemory {
...
             cross_session: self.cross_session,
+            // TODO(pin bump): decide this deliberately before shipping the
+            // engine bump that introduced the field. `None` reproduces exactly
+            // what this call did before it existed — the engine falls back to
+            // its ambient task-local. But the engine's own comment says that
+            // ambient reads `None` on the far side of the module boundary,
+            // because a `cdylib` has its own statics, and this call goes
+            // through the module. So `None` here is bug-compatible, not
+            // correct: it hands the agent back what it just said.
+            exclude_session_id: None,
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/openhuman/agent/tinyagents/host/agent_memory.rs]: Agreed, and applied — with one correction to the second half that is worth recording, because taking it literally broke real contract.

**The production change is right, and my TODO was wrong.** I had left `None` reasoning that `session_id` already scopes to a session, so excluding the same one would empty the result — especially since `cross_session` defaults to `false`. The engine's own doc on `query_namespace_hits_excluding_session` settles it the other way:

> Only documents are session-filtered — KV rows carry no session concept, and episodic/event hits already have their own dedicated session-scoping (`RecallOpts::session_id` / `cross_session`).

Different tiers, so scoping to a session and excluding it is not a contradiction. That doc also names this exact situation: the harness saves the user's message as a `[conversation]` document tagged with the active thread before the agent runs, so a recall issued during that turn can retrieve its own trigger. Set to the active thread, with the reasoning at the call site instead of a TODO.

**The test suggestion needed adjusting.** "Update the test backend to honor this field and assert that matching-session entries are excluded" is the obvious reading, and doing exactly that broke two existing tests — `a_thread_hint_narrows_scope_and_keeps_unscoped_rows` and `cross_session_with_a_thread_hint_recalls_other_sessions`. Both require the **same-session** row to come back, because a thread hint narrows *to* a session rather than away from it.

They are right and the stub change was wrong. `StubMemory` is a flat row list with `category: Conversation` on every row, so it cannot represent the document-versus-KV split the real filter turns on; applying the exclusion to every row asserts a backend behaviour that does not exist.

So the stub records which exclusion was asked for, and the test asserts the request: `Some("t1")` with a thread hint, `None` without. Asking for the right exclusion is the host's job; applying it is the engine's, and the engine's suite has the tiers to prove it against.

**Comment by YellowSnnowmann** [comment]
Correct on both counts, and both are mine. `store_stats` propagates its driver error the same way, and `pipeline_status_rpc` does not degrade either — so the comment's contrast was wrong in both directions.

Rewritten close to your suggested text: a driver error propagates, the same way `store_stats` propagates its own; `backfill_status_rpc` is the case that matters, because it is asked whether a modal may close and guessing "nothing pending" would dismiss it over a live backfill; and a driver that does not serve `Maintenance` still reports empty rather than erroring, because it has no queue to be behind on.

**Comment by YellowSnnowmann** [comment]
Two things about the local suite, recorded so a red run is not re-diagnosed from scratch.

**`budget_gate::{dropping_the_crate_permit_releases_the_scheduler_permit, explicit_release_returns_capacity_before_end_of_scope}`** — semaphore timing. They pass in isolation and fail only under full-suite parallel load. Nothing in this PR touches that module.

**`memory::{tree_e2e_tests, sync_pipeline_e2e_tests}`** — these can *hang* rather than fail in a full local run. `sample` puts them in `modules::host::runtime` (`host.rs:129`) with the `openhuman-module-bus` thread live, which is the documented one-runtime-per-process trap: tinybus binds its broker tasks to the runtime that created them, the module is loaded once per process and never unloaded, so a later test finds a dead broker and blocks. Three of them run in one process.

Checked rather than assumed, because "the suite hung" is not evidence of anything on its own:

- `git diff --name-only` — this PR touches none of those files.
- Run alone, `tree_e2e_tests` is `2 passed` in 1.08s.
- An earlier full run of this same code finished `11557 passed`.

What is verified on `dd0798e9`:

```
cargo check --all-targets --features "$(bash scripts/ci/product-features.sh)"   # 0 errors
cargo clippy -p openhuman --features "$(bash scripts/ci/product-features.sh)" -- -D warnings   # 0
cargo clippy -p openhuman -- -D warnings                                        # 0 (contributor default set)
cargo fmt --all -- --check                                                      # clean
cargo metadata --locked                    # root ok
cargo metadata --locked --manifest-path app/src-tauri/Cargo.toml                # shell ok
```

Both clippy invocations matter separately — CI runs the product feature set and the contributor default set, and code can compile under one and not the other.

File: src/openhuman/memory/tree/tree/rpc.rs [hunk 2]
```diff
@@ -251,6 +250,67 @@ pub async fn get_chunk_rpc(
...
+    let Some(maintenance) = binding.provider().as_maintenance() else {
+        log::debug!(
+            "[memory-tree][rpc] store_stats: driver '{}' does not serve Maintenance; reporting empty",
+            binding.driver_id()
+        );
+        return Ok(Default::default());
+    };
+    maintenance
+        .store_stats()
+        .await
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/openhuman/memory/tree/tree/rpc.rs:285]: Confirmed, and this is the blocker on the PR — thank you.

Verified both halves. `ModuleMemoryProvider` forwarded only `reembed`/`compact`/`consolidate`/`doctor`, so the four new members fell through to the contract defaults. Added the forwarders (`RetryFailed`, `StoreStats`, `QueueStats`, `LatestQueueFailure`).

The pin is the part that cannot be fixed inside this PR. Merging `main` brought the registry from v1.0.1 to **v1.2.0**, but v1.2.0 was cut at 11:03 today and both contract PRs merged after it — tinyhumansai/tinymemory#85 (the diagnostics) and #86 (`retry_failed`). So no published artifact implements these members yet, and `git tag --contains` on that work returns nothing.

With forwarders the failure mode changes from silent to loud — the bus answers `UnknownMethod` instead of the members quietly returning empty — but `pipeline_status`, `backfill_status` and the retry remediation are broken either way against a v1.2.0 module.

So this needs, in order:

1. a tinymemory release containing #85 and #86 (manual `workflow_dispatch`, so a maintainer action);
2. `modules/registry.rs` re-pinned to it with the published SHA-256 digests;
3. then this PR.

I have flagged it as do-not-merge-yet rather than leaving the ordering implicit. Repointing `vendor/tinymemory` back to v1.2.0 to match the registry is not an option that helps: v1.2.0 lacks the members at source level too, so it would just move the failure to compile time.

File: src/openhuman/memory/tree/tree/rpc.rs [hunk 3]
```diff
@@ -272,29 +332,26 @@ pub async fn backfill_status_rpc(
...
+            msg
+        })?;
+    // Ready + running, not `total - done`: a failed backfill job is finished
+    // with, and counting it as pending leaves the modal open forever.
+    let pending_jobs: u64 = queue.ready + queue.running;
+    // Still the engine's process-global. It covers the instant between one
+    // backfill link settling and the next being enqueued, which the counts
+    // cannot see — but it is scoped to the process, not to this store, and a
+    // second memory subtree running its own backfill would answer `true` here.
+    // Putting it on `QueueStats` would have made a per-store API promise that
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/openhuman/memory/tree/tree/rpc.rs]: Agreed, and the reasoning matters because it cuts against an earlier review on the upstream PR.

`backfill_in_progress()` was briefly a field on `QueueStats` in tinyhumansai/tinymemory#86. Review there objected — correctly — that it is a process-global while `open_store` serves one store per memory subtree, so a per-store snapshot promising per-store scoping is worse than an obviously global call. I dropped the field and left the host reading the host-linked static, with a comment saying why.

Your point is the other half of the same problem: once re-embedding runs in the module, that host-linked static is no longer merely coarse, it is *always false*. The frontend poll then closes the modal during exactly the gap the flag exists to cover.

Both objections are satisfied by the same shape, which is neither of the two I have tried: a standalone `Maintenance` member rather than a `QueueStats` field, documented as driver-process-wide and explicitly not store-scoped. That keeps the honest scope in the signature instead of implying one it does not have, and it comes from the bound driver so it reflects where the backfill actually runs.

That is a third contract member and therefore rides the same release this PR is already blocked on, so I am doing it upstream rather than patching around it here. Until it lands, this call site keeps the host-side static with a comment recording that it under-reports on the module path — which is the pre-existing behaviour on this branch, not something the migration introduced.

**Comment by YellowSnnowmann** [comment]
## ⛔ Do not merge yet — needs a TinyMemory release first

Codex caught this and it is correct. Recording it at top level so the ordering is not buried in a thread.

The desktop path reaches memory through the compiled module, and `modules/registry.rs` pins a **published artifact**. That pin is now v1.2.0 (from the `main` merge, previously v1.0.1). v1.2.0 was cut at 11:03 today; both contract PRs merged after it:

- tinyhumansai/tinymemory#85 — `store_stats`, `queue_stats`, `latest_queue_failure`
- tinyhumansai/tinymemory#86 — `retry_failed`

`git tag --contains` on that work returns nothing, so **no release implements these members**. Against a v1.2.0 module the four calls fail, and `pipeline_status`, `backfill_status` and the "Retry failed" control break with them.

### Required order

1. **Cut a TinyMemory release** containing #85 and #86 — manual `workflow_dispatch` on `release.yml`, minor bump.
2. **Re-pin `modules/registry.rs`** to that tag with the published per-platform SHA-256 digests (taken verbatim from the release, not recomputed locally).
3. **Then merge this PR.**

### What changed here in response

`ModuleMemoryProvider` now forwards all four members (`RetryFailed`, `StoreStats`, `QueueStats`, `LatestQueueFailure`); it previously stopped at `doctor`. That turns the failure from silent to loud — the bus answers `UnknownMethod` rather than the members quietly returning contract defaults — but it does not make the PR safe on its own. Step 1 does.

Re-pointing `vendor/tinymemory` back to v1.2.0 to match the registry does not help: v1.2.0 lacks the members at source level too, so it moves the failure from run time to compile time.

Separately: `backfill_in_progress` needs a `Maintenance` member of its own (Codex P2) and rides the same release. Explanation in that thread.

File: src/openhuman/modules/memory.rs [hunk 4]
```diff
@@ -969,6 +969,18 @@ impl MemoryMaintenance for ModuleMemoryProvider {
...
+    }
+    async fn store_stats(&self) -> Result<StoreStats, MemoryError> {
+        module_call!(self, "store_stats", "StoreStats", ())
+    }
+    async fn queue_stats(&self, kind: Option<&str>) -> Result<QueueStats, MemoryError> {
+        module_call!(self, "queue_stats", "QueueStats", (kind,))
+    }
+    async fn latest_queue_failure(&self) -> Result<Option<QueueFailure>, MemoryError> {
+        module_call!(self, "latest_queue_failure", "LatestQueueFailure", ())
+    }
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/openhuman/modules/memory.rs:987]: Checked this against the tree and I'm not applying it — `classify_sdk_error` does not exist in this repository (`grep -rn 'classify_sdk_error' src/` returns nothing), and there is no backend-call header convention on this seam. I think the guideline being matched here belongs to an HTTP/SDK client rather than to a TinyBus proxy call.

The mapping these five methods use is deliberate, and it is the one thing on this seam that must not be re-implemented locally. `module_call!` routes every failure through `from_bus`, which is the **shared** table — the same one the module maps *out* through. `modules/memory.rs` records why:

> reimplementing the mapping here is what would let a `PathEscape` arrive as an `Invalid`, silently reclassifying a sandbox escape as a caller mistake.

All ~90 forwarders on `ModuleMemoryProvider` go through that single macro, including the four you flagged and `backfill_in_progress` alongside them. Giving these five a different error path would make them the only members whose bus errors are classified by host-side code, which is the divergence the shared table exists to prevent.

Happy to reconsider if `classify_sdk_error` is meant to be introduced by another PR — but as things stand there is nothing to route through.

File: vendor/tinyflows [hunk 5]
```diff
@@ -1 +1 @@
-Subproject commit 4e2e04267ae62f04a3f8faccaf0bdad9f17a9c05
+Subproject commit 8f63a2734f13bb5b7b974615bb8d4abd2df57894
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply vendor/tinyflows]: Good catch — this was real, and it was a silent revert rather than an intentional pin.

The branch carried `vendor/tinyflows` at `8f63a273`, while `main` had moved to `4e2e0426` (`048762c66`) and has since moved again to `7ba0b918` (#5716). Nothing in this PR has any business pinning tinyflows; the older gitlink rode in through a merge.

The reason it slipped past the previous merge is worth recording: the two submodule commits are not ancestor-related from git's point of view at merge time, so `git merge` reports

```
Failed to merge submodule vendor/tinyflows (commits don't follow merge-base)
CONFLICT (submodule): Merge conflict in vendor/tinyflows
```

rather than fast-forwarding the path. A merge that resolves that quietly the wrong way leaves exactly this — a branch that reverts someone else's bump while showing a clean diff against its own base.

Fixed in 2d97fc434: merged `upstream/main` and resolved the submodule onto `7ba0b918`, which I verified is strictly newer (`git -C vendor/tinyflows merge-base --is-ancestor 8f63a273 7ba0b918` succeeds). The gitlink now matches `upstream/main` byte for byte.

**Comment by YellowSnnowmann** [comment]
Applied in 98ba5ac75 — you were right that the assertion proved nothing.

The request sets five filters, but the store held only the matching chunk, so a handler that dropped every filter still returns exactly one row and `assert_eq!(listed.len(), 1)` passes.

There are now three decoys — wrong source id, outside the time window, wrong owner — and the assertion compares the returned ids to exactly `[chunk.id]` rather than counting them.

I mutation-checked it rather than trusting that it discriminates: replacing `owner: Some("round18-user")` with `None` makes it fail, and the failure message names the decoy that leaked through.

```
assertion `left == right` failed: every filter must discriminate:
0ef310c4… (source id), 70978db1… (window), a6ee6427… (owner) are all in the store
```

Source kind is the one filter still not covered by a decoy — `chunk_id` derives from the kind, and a non-chat chunk would need a different seeder. Left it out rather than adding a fourth shape for the weakest of the five.

File: src/openhuman/memory/tree/tree/rpc.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/openhuman/memory/tree/tree/rpc.rs:285]: Resolved — the release exists now.

tinymemory **v1.3.0** was cut and published today, and it is the first release carrying all four members this PR calls: `store_stats`, `queue_stats`, `latest_queue_failure` (tinyhumansai/tinymemory#85) and `retry_failed` (#86), plus `backfill_in_progress` (#89) for your P2 below.

`registry.rs` is re-pinned to it in d73f5961a, with the digests for all eleven platform archives taken verbatim from the release's own `checksum.toml` — never recomputed locally, since a recomputed digest agrees with itself no matter what was served. I verified every one of the eleven against the published manifest: no mismatches, nothing missing, all `1.3.0` archives.

`vendor/tinymemory` moved to the release commit alongside it, and both lockfiles moved with it because the facade version went 1.2.0 → 1.3.0.

The do-not-merge note at the top of this PR no longer applies.

File: src/openhuman/memory/tree/tree/rpc.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/openhuman/memory/tree/tree/rpc.rs]: Resolved in 96b2eb0f3, upstream first as promised.

tinyhumansai/tinymemory#89 added `backfill_in_progress` as a standalone `Maintenance` member rather than a `QueueStats` field, documented as driver-process-wide and explicitly not store-scoped — which satisfies both this finding and the objection that got it dropped from #86.

Host side: a `module_call!` forwarder in `modules/memory.rs`, a `backfill_in_progress(config)` helper beside `queue_stats`, and `backfill_status_rpc` now reading the driver instead of `tinymemory_core::queue::backfill_in_progress()`. A read failure degrades to the counts rather than failing a polled RPC.

It is pinned by a test rather than left to inspection: `backfill_status_reports_the_drivers_flag_when_the_counts_are_empty` asserts exactly the instant you described — zero ready, zero running, and the driver still reporting a backfill. `FixedDiagnostics` grew a `backfilling()` builder so the flag can be set independently of the counts, since the whole point is that it is not derivable from them.

One doc note went with it: the sibling test said `in_progress` could not be asserted for the empty case because the flag was a process-global shared across parallel tests. It comes from the bound driver now, so it can be, and is.

File: src/openhuman/memory/tools/store.rs [hunk 6]
```diff
@@ -168,6 +169,26 @@ mod tests {
...
+    /// landed where a caller would look for it.
+    async fn stored(namespace: &str, key: &str) -> Option<MemoryEntry> {
+        active_memory_guard()
+            .await
+            .expect("a bound memory guard")
+            .get(namespace, key)
+            .await
+            .expect("read back through the guard")
+    }
+
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/openhuman/memory/tools/store.rs:191]: Applied in 933ab148d — and you found the sharper half of it. The three absence assertions were not merely reading the wrong store, they were **vacuous**: `assert!(mem.get(...).is_none())` against a handle the tool never writes to holds whether or not the refusal worked. That is worse than no assertion, because it reads as coverage.

All three now go through `stored(...)`.

One thing beyond what you flagged: `store_blocked_in_readonly_mode` and `store_blocked_when_rate_limited` both used `global/lang`, which is also `store_core`'s key. Harmless while every module test gets its own process, and a latent trap the moment one does not — the absence each asserts could be satisfied by the other's leftover. They have distinct keys now (`readonly_lang`, `ratelimited_lang`).

Verified against a real cdylib, one process per test: `store_core`, `store_with_custom_category`, `store_strips_custom_prefix_from_wire_category`, `store_rejects_secret_like_content`, `store_blocked_in_readonly_mode`, `store_blocked_when_rate_limited` — all six pass.

File: tests/raw_coverage/memory_tree_sync_deep_raw_coverage_e2e.rs [hunk 7]
```diff
@@ -556,20 +494,48 @@ async fn memory_tree_rpc_status_set_enabled_backfill_and_ingest_errors() {
...
+                path_scope: None,
+            },
+            token_count: 32,
+            seq_in_source: 3,
+            created_at: ts,
+            partial_message: false,
+        };
+        upsert_chunks(&cfg, std::slice::from_ref(&chunk)).expect("upsert decoy");
+        chunk
+    };
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply tests/raw_coverage/memory_tree_sync_deep_raw_coverage_e2e.rs:538]: Fair — I had left it out and said so, and you are right that the reasoning did not hold. Applied in 933ab148d.

There is now an `Email` decoy matching the request on source id, owner and the time window, differing only in kind, and its id is in the assertion message alongside the other three.

My stated reason for skipping it was that `chunk_id` derives from the source kind so it needs a bespoke seeder rather than `sample_chunk`. That part was true but irrelevant — the `wrong_owner` decoy already builds a `Chunk` literal inline, so the fourth costs the same as the third did.

Mutation-checked rather than assumed: replacing `source_kind: Some("chat")` with `None` fails the test and names the leaked decoy.

```
assertion `left == right` failed: every filter must discriminate:
0ef310c4… (source id), 70978db1… (window), a6ee6427… (owner), c3d12740… (source kind)
are all in the store
```

All five filters in the request are now covered by a decoy each.
